### DISCUSSION TITLE
How to combine, merge, or otherwise edit Companion entries?

### DISCUSSION THREAD

**Discussion by wanglophile** [simolus3/drift#3154] [Q&A] [answered]
Can I combine, merge or otherwise edit Companion entries?

For example, given an initial Companion entry of

```
final ItemsCompanion data = ItemsCompanion(
  foo: 'foo',
  bar: 'bar',
);
```

How can I add a `baz: 'baz'`?

Something like:

```
Future <int> createItemWithBaz(ItemsCompanion entry) {
  // This obviously does not work
  final ItemsCompanion entryWithBaz = ItemsCompanion(
    baz: 'baz',
  ).merge(entry);

  return into(items).insert(entryWithBaz);
}
```

with the end goal of having a new Companion of

```
ItemsCompanion(
  foo: 'foo',
  bar: 'bar',
  baz: 'baz',
);
```

**Comment by simolus3** [comment, marked as answer]
I didn't try it, but I think this should work:

```dart
extension Merge<T> on Insertable<T> {
  Insertable<T> merge(Insertable<T> other) {
    return _MergeInsertable(this, other);
  }
}

final class _MergeInsertable<T> implements Insertable<T> {
  final Insertable<T> _base;
  final Insertable<T> _mergedIn;

  _MergeInsertable(this._base, this._mergedIn);

  @override
  Map<String, Expression<Object>> toColumns(bool nullToAbsent) {
    return {
      ..._base.toColumns(nullToAbsent),
      ..._mergedIn.toColumns(nullToAbsent),
    };
  }
}
```

I'm not entirely convinced that this should be a part of drift though. We have `copyWith`, but I don't think merges are used that often.

**Reply by wanglophile** [reply]
Thank you!

**Reply by wanglophile** [reply]
Follow-up question: how can I see if if a property exists in a Companion entry?

So in my original example, if `data` has a `baz` property in it?

**Reply by simolus3** [reply]
> Follow-up question: how can I see if if a property exists in a Companion entry?

The only way at the moment is to call `toColumns(true)` and then check whether the map contains the key for the column you're interested in.
